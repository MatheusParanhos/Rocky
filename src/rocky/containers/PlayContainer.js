import React, { Component } from "react";
import { Text, View, StyleSheet } from "react-native";
import { Button } from "../../shared/components/Button";
import Scissors from "../components/ScissorsButton";
import Paper from "../components/PaperButton";
import Rock from "../components/RockButton";
import { deviceWidth } from "../../theme/theme";
import Buttons from "../components/Buttons";
import Result from "../components/Result";
export default class PlayContainer extends Component {
  constructor(props) {
    super(props);
    // Initialize states
    this.state = {
      userChoice: "",
      computerChoice: "",
      result: "",
      countdown: ""
    };
  }

  // componentDidMount() {
  //   this.JokenPo("rock");
  // }
  //** JokenPo function
  JokenPo(userChoice) {
    // Reinitialize state
    this.setState({
      result: ""
    });
    // Initialize variables
    let result = "";
    let computerChoice = "";

    // Generate a random number
    const randomNumber = Math.floor(Math.random() * 3);

    switch (randomNumber) {
      case 0:
        computerChoice = "rock";
        break;
      case 1:
        computerChoice = "paper";
        break;
      case 2:
        computerChoice = "scissors";
        break;
    }
    // Initialize Countdown
    this.setState({
      countdown: 3
    });

    startCountdown = () => {
      this.handleClock = setInterval(() => {
        decrementClock();
      }, 1000);
    };

    decrementClock = () => {
      if (this.state.countdown === 1) clearInterval(this.handleClock);
      this.setState(prevState => ({ countdown: prevState.countdown - 1 }));
      if (this.state.countdown === 0) this.setState({ countdown: "" });
    };

    // Start Countdown
    startCountdown();

    // Compare answers
    let compare = function(computerChoice, userChoice) {
      if (computerChoice === userChoice) {
        result = "Fair enough, it's a draw!";
      } else if (computerChoice === "scissors" && userChoice === "paper") {
        result = "Oh, no you lost!";
      } else if (computerChoice === "rock" && userChoice === "scissors") {
        result = "Oh, no you lost!";
      } else if (computerChoice === "paper" && userChoice === "rock") {
        result = "Oh, no you lost!";
      } else {
        result = "Nice, you won!";
      }
    };
    // Call compare function
    compare(computerChoice, userChoice);

    // Set state of result
    this.countdown = setTimeout(() => {
      this.setState({
        userChoice,
        computerChoice,
        result
      });
    }, 3000);
  }

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.top} />
        <View style={styles.bottom}>
          <Result
            choice={this.state.computerChoice}
            timer={this.state.countdown}
            result={this.state.result}
          />

          <Buttons
            {...this.props}
            onScissorsPress={() => {
              this.JokenPo("scissors");
            }}
            onPaperPress={() => {
              this.JokenPo("paper");
            }}
            onRockPress={() => {
              this.JokenPo("rock");
            }}
          />
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: deviceWidth,
    backgroundColor: "white"
  },
  top: {
    flex: 1
    // backgroundColor: 'red'
  },
  bottom: {
    flex: 1,
    justifyContent: "flex-end"
    // width:'100%',
    //backgroundColor: 'blue'
  }
});
