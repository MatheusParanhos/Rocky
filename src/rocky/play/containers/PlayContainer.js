import React, { Component } from "react";
import { View, StyleSheet } from "react-native";
import { deviceWidth } from "../../../theme/theme";
import Buttons from "../components/Buttons";
import Result from "../components/Result";
import PlayAgain from "../components/PlayAgain";
export default class PlayContainer extends Component {
  constructor(props) {
    super(props);
    // Initialize states
    this.state = {
      userChoice: "",
      computerChoice: "",
      result: "",
      countdown: null,
      isPlaying: false,
      firstTime: true
    };
  }

  /**
   * Reset previous state
   * @param {string} userChoice
   */
  resetState(userChoice) {
    // Reinitialize state
    this.setState(
      {
        result: "",
        userChoice: "",
        computerChoice: "",
        isPlaying: true
      },
      () => this.generateComputerChoice(userChoice)
    );
  }
  /**
   * Generate computer choice function
   * @param {string} userChoice
   */
  generateComputerChoice(userChoice) {
    // Initialize variables
    let computerChoice = "";

    // Generate a random number
    const randomNumber = Math.floor(Math.random() * 3);
    switch (randomNumber) {
      case 0:
        computerChoice = "rock";
        break;
      case 1:
        computerChoice = "paper";
        break;
      case 2:
        computerChoice = "scissors";
        break;
    }
    this.jokenPo(userChoice, computerChoice);
  }
  /**
   * Joken Po function
   * @param {string} userChoice
   * @param {string} computerChoice
   * 
   * Here we define a constraint that will dictate rules for the winner
     This way we can achieve the so called 'extensibility' of code,
     becoming easier to scale into more/different set of rules
   */
  jokenPo(userChoice, computerChoice) {
    let constraint = { paper: "rock", scissors: "paper", rock: "scissors" };
    if (userChoice === constraint[computerChoice]) {
      result = "Oh no, you lost!";
    } else {
      result = "Nice, you won!!";
    }
    if (computerChoice === userChoice) {
      result = "Fair enough, it's a draw!";
    }
    // Call setCountdown function
    this.setCountDown(userChoice, computerChoice, result);
  }

  /**
   * Set Countdown
   * @param {string} userChoice
   * @param {string} computerChoice
   * @param {string} result
   */
  setCountDown(userChoice, computerChoice, result) {
    // Initialize Countdown
    this.setState(
      {
        countdown: 3
      },
      () => this.countdownAndShowResults(userChoice, computerChoice, result)
    );
  }

  countdownAndShowResults(userChoice, computerChoice, result) {
    /**
     * Timer Functions
     * @param
     */
    startCountdown = () => {
      this.handleClock = setInterval(() => {
        decrementClock();
      }, 1000);
    };
    // Start Countdown
    startCountdown();

    // Decrement clock states to avoid unwanted behaviours
    decrementClock = () => {
      if (this.state.countdown === 1) clearInterval(this.handleClock);
      this.setState(prevState => ({ countdown: prevState.countdown - 1 }));
      if (this.state.countdown === 0) this.setState({ countdown: null });
      if (this.state.countdown < 0) this.setState({ countdown: null });
    };

    // Set state of result
    this.countdown = setTimeout(() => {
      this.setState({
        userChoice,
        computerChoice,
        result,
        isPlaying: false,
        firstTime: false
      });
    }, 3000);
  }

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.top} />
        <View style={styles.bottom}>
          <Result
            computerChoice={this.state.computerChoice}
            userChoice={this.state.userChoice}
            countdown={this.state.countdown}
            result={this.state.result}
          />
          {this.state.firstTime === false &&
            this.state.isPlaying === false && <PlayAgain />}
          {this.state.isPlaying === false && (
            <Buttons
              {...this.props}
              onScissorsPress={() => {
                this.resetState("scissors");
              }}
              onPaperPress={() => {
                this.resetState("paper");
              }}
              onRockPress={() => {
                this.resetState("rock");
              }}
            />
          )}
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: deviceWidth,
    backgroundColor: "white"
  },
  top: {
    flex: 1
  },
  bottom: {
    flex: 1,
    justifyContent: "flex-end"
  }
});
