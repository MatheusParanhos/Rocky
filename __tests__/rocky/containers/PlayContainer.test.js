import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import PlayContainer from '../../../src/rocky/play/containers/PlayContainer'

Enzyme.configure({ adapter: new Adapter() });
var wrapper;

test('Render matches snapshot', () => {
    wrapper = Enzyme.shallow(<PlayContainer />);
  expect(wrapper).toMatchSnapshot();
})
