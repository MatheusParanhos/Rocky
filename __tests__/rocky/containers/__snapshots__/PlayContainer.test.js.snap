// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlayContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Component
          style={
            Object {
              "flex": 1,
            }
          }
        />,
        <Component
          style={
            Object {
              "flex": 1,
              "justifyContent": "flex-end",
            }
          }
        >
          <Result
            computerChoice=""
            countdown={null}
            result=""
            userChoice=""
          />
          <Buttons
            computerChoice=""
            onPaperPress={[Function]}
            onRockPress={[Function]}
            onScissorsPress={[Function]}
            result=""
            timer={0}
            userChoice=""
          />
        </Component>,
      ],
      "style": Object {
        "backgroundColor": "white",
        "flex": 1,
        "width": 750,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "style": Object {
            "flex": 1,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Result
              computerChoice=""
              countdown={null}
              result=""
              userChoice=""
            />,
            <Buttons
              computerChoice=""
              onPaperPress={[Function]}
              onRockPress={[Function]}
              onScissorsPress={[Function]}
              result=""
              timer={0}
              userChoice=""
            />,
          ],
          "style": Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "computerChoice": "",
              "countdown": null,
              "result": "",
              "userChoice": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "computerChoice": "",
              "onPaperPress": [Function],
              "onRockPress": [Function],
              "onScissorsPress": [Function],
              "result": "",
              "timer": 0,
              "userChoice": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Component
            style={
              Object {
                "flex": 1,
              }
            }
          />,
          <Component
            style={
              Object {
                "flex": 1,
                "justifyContent": "flex-end",
              }
            }
          >
            <Result
              computerChoice=""
              countdown={null}
              result=""
              userChoice=""
            />
            <Buttons
              computerChoice=""
              onPaperPress={[Function]}
              onRockPress={[Function]}
              onScissorsPress={[Function]}
              result=""
              timer={0}
              userChoice=""
            />
          </Component>,
        ],
        "style": Object {
          "backgroundColor": "white",
          "flex": 1,
          "width": 750,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "style": Object {
              "flex": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Result
                computerChoice=""
                countdown={null}
                result=""
                userChoice=""
              />,
              <Buttons
                computerChoice=""
                onPaperPress={[Function]}
                onRockPress={[Function]}
                onScissorsPress={[Function]}
                result=""
                timer={0}
                userChoice=""
              />,
            ],
            "style": Object {
              "flex": 1,
              "justifyContent": "flex-end",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "computerChoice": "",
                "countdown": null,
                "result": "",
                "userChoice": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "computerChoice": "",
                "onPaperPress": [Function],
                "onRockPress": [Function],
                "onScissorsPress": [Function],
                "result": "",
                "timer": 0,
                "userChoice": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
